File: clases/clase05/conjuntos_finitos/ejercicio_2_1_a_Python/ejercicio_2_1_a.py
def ConjuntoVacio():
    return []

def ConjuntoUnitario(elemento):
    return [elemento]

def EsConjuntoVacio(conjunto):
    return conjunto == []

def Pertenece(elemento, conjunto):
    return elemento in conjunto

def Agregar(elemento, conjunto):
    conjunto.append(elemento)
    return conjunto

def Quitar(elemento, conjunto):
    nuevo = [item for item in conjunto if item != elemento]
    return nuevo

def Union(conjunto1, conjunto2):
    resultado = []
    for elemento in conjunto1:
        resultado.append(elemento)
    for elemento in conjunto2:
        resultado.append(elemento)
    return resultado

def Interseccion(conjunto1, conjunto2):
    resultado = []
    for actual in conjunto1:
        if Pertenece(actual, conjunto2):
            resultado.append(actual)
    return resultado

def Diferencia(conjunto1, conjunto2):
    resultado = []
    for actual in conjunto1:
        if not Pertenece(actual, conjunto2):
            resultado.append(actual)
    return resultado

def Cardinalidad(conjunto):
    conjunto = list(set(conjunto))
    return len(conjunto)

def main():
    conjunto1 = ConjuntoVacio()
    assert conjunto1 == [], "Error de la funcion ConjuntoVacio."

    conjunto2 = ConjuntoUnitario(13);
    conjunto3 = ConjuntoUnitario(25);
    assert conjunto2 == [13] and conjunto3 == [25], "Error de la funcion ConjuntoUnitario."

    assert EsConjuntoVacio(conjunto1) and not EsConjuntoVacio(conjunto2), "Error de la funcion EsConjuntoVacio."

    assert Pertenece(13, conjunto2) and not Pertenece(13, conjunto3), "Error de la funcion Pertenece."

    assert Agregar(15, conjunto2) == [13, 15], "Error de la funcion Agregar."
    assert Agregar(17, conjunto2) == [13, 15, 17], "Error de la funcion Agregar."
    assert Agregar(15, conjunto2) == [13, 15, 17, 15], "Error de la funcion Agregar."

    conjunto2 = Quitar(15, conjunto2)
    assert conjunto2 == [13, 17], "Error de la funcion Quitar."
    conjunto2 = Quitar(15, conjunto2)

    conjunto2 = Union(conjunto2, conjunto3)
    assert conjunto2 == [13, 17, 25], "Error de la funcion Union."

    conjunto2 = Interseccion(conjunto2, conjunto3)
    assert conjunto2 == [25], "Error de la funcion Interseccion."

    conjunto2 = Agregar(11, conjunto2)
    conjunto2 = Diferencia(conjunto2, conjunto3)
    assert conjunto2 == [11], "Error de la funcion Diferencia."

    conjunto2 = Agregar(11, conjunto2)
    assert Cardinalidad(conjunto2) == 1, "Error de la funcion Cardinalidad."

    print("Todos los casos estan OK,")

if __name__ == "__main__":
    main()

File: clases/clase-05-punteros-y-conjuntos-finitos/conjuntos_finitos/C/ejercicio_2_1_a.c
/*
 * Implementacion del TAD de los conjuntos finitos utilizando como
 * representacion del tipo Conjunto como vectores de Elementos
 */

#include <stdio.h>
#include <stdlib.h>
#include <assert.h>
#include "ejercicio_2_1_a.h"

// (tamanio del arreglo, arreglo)
int main(int argc, char* argv[]) {
  // Si al invocar el programa no me dieron el tamanio, uso 10^4 por defecto
  if (argc == 1) {
    MAX_N = 10000;
  } else {
    MAX_N = atoi(argv[1]);
  }
  printf("El conjunto puede tener hasta %d elementos.\n", MAX_N);

  Conjunto conjunto1 = ConjuntoVacio();
  assert(EsConjuntoVacio(conjunto1));

  Conjunto conjunto2 = ConjuntoUnitario(69);
  assert(!EsConjuntoVacio(conjunto2));
  assert(*(conjunto2.contenido) == 69);
  assert(Pertenece(69, conjunto2));
  //assert(!Pertenece(69, conjunto2));
  
  Agregar(6, &conjunto1);
  Agregar(6, &conjunto1);
  Quitar(6, &conjunto1);
  assert(EsConjuntoVacio(conjunto1));

  Agregar(6, &conjunto2);
  Agregar(69, &conjunto2);
  Quitar(69, &conjunto2);
  assert(Pertenece(6, conjunto2));
  assert(!Pertenece(69, conjunto2));

  Conjunto conjunto3 = Diferencia(conjunto2, conjunto1);
  assert(Pertenece(6, conjunto3));
  free(conjunto3.contenido);

  Agregar(69, &conjunto1);
  conjunto3 = Union(conjunto1, conjunto2);
  assert(Pertenece(6, conjunto3));
  assert(Pertenece(69, conjunto3));
  free(conjunto3.contenido);

  Agregar(6, &conjunto1);
  conjunto3 = Interseccion(conjunto1, conjunto2);
  assert(Pertenece(6, conjunto3));
  assert(!Pertenece(69, conjunto3));
  free(conjunto3.contenido);

  Agregar(69, &conjunto1);
  assert(Cardinalidad(&conjunto1) == 2);
  assert(Pertenece(6, conjunto1));
  assert(Pertenece(69, conjunto1));
  assert(Cardinalidad(&conjunto2) == 1);

  free(conjunto1.contenido);
  free(conjunto2.contenido);
  printf("Todos los casos estan OK.\n");
  return 0;
}
